controlplane/kubeadm/internal/control_plane.go

```diff
@@ -205,17 +200,124 @@ func (c *ControlPlane) NewMachine(infraRef, bootstrapRef *corev1.ObjectReference
 	}
 }
 
-// NeedsReplacementNode determines if the control plane needs to create a replacement node during upgrade.
-func (c *ControlPlane) NeedsReplacementNode() bool {
-	// Can't do anything with an unknown number of desired replicas.
-	if c.KCP.Spec.Replicas == nil {
-		return false
-	}
-	// if the number of existing machines is exactly 1 > than the number of replicas.
-	return len(c.Machines)+1 == int(*c.KCP.Spec.Replicas)
+func (c *ControlPlane) DeletingMachines() FilterableMachineCollection {
+	return c.Machines.Filter(machinefilters.HasDeletionTimestamp)
 }
 
 // HasDeletingMachine returns true if any machine in the control plane is in the process of being deleted.
 func (c *ControlPlane) HasDeletingMachine() bool {
-	return len(c.Machines.Filter(machinefilters.HasDeletionTimestamp)) > 0
+	return c.DeletingMachines().Any()
+}
+
+// UnhealthyMachines returns the machines that need remediation.  If cluster
+// size is less than 3, will return an empty list.
+func (c *ControlPlane) UnhealthyMachines() FilterableMachineCollection {
```

Comment:
Does there need to be a differentiation between unhealthy machines and machines that need remediation (I'm thinking where you could have a failed machine that does not have the owner remediation annotation)?

Reply:
I don't know. This relates to the prior comment, what do you suggest we might want to do with those machines? We could reimplement some facets of MHC and remediate machines without the remediation condition, but it's not clear to me that's a good idea. We could also just freeze reconciliation or something.

Reply:
+1 to not attempting to replicate MHC for handling.

I think freezing reconciliation would make sense, since it would still allow for external remediation, but would also prevent us from taking additional actions that could potentially further degrade the control plane

Reply:
I opened #3230 for this